File: lab1/cs/BreakThread.cs
public class BreakThread {
	private static List<ComputingThreadStatus> associatedThreads = [];
	
	public BreakThread (int amountThreads) {
    ComputingThread[] threads = new ComputingThread[amountThreads];
    int min = 1;
    int max = 10;
    Random rnd = new Random();

    for (int i = 0; i < amountThreads; i++){
      int step = rnd.Next(max - min + 1) + min;
      int runDuration = rnd.Next(max - min + 1) + min;
      threads[i] = new ComputingThread(i, 1, step, runDuration);
    }

		var threadStatuses = threads.Select(thread => new ComputingThreadStatus(thread)).ToList();
		threadStatuses.Sort((t1, t2) => t1.thread.runDuration.CompareTo(t2.thread.runDuration));
    
		associatedThreads = threadStatuses.Select((threadStatus, idx) => {
      threadStatus.thread.queueNumber = idx;
      return threadStatus;
    }).ToList();
	}
	
	public static bool canBreak (int queueNumber) {
		return associatedThreads[queueNumber].canBreak;
	}
	
	public void Start () {
		List<DateTime> endTimestamps = associatedThreads.Select(threadStatus => {
			return DateTime.Now.Add(TimeSpan.FromSeconds(threadStatus.thread.runDuration));
		}).ToList();

    for (int i = 0; i < endTimestamps.Count; i++){
      new Thread(associatedThreads[i].thread.Start).Start();
    }
		
		int idx = 0;
		while (DateTime.Now < endTimestamps[associatedThreads.Count - 1]) {
			int sleepTime = (int)Math.Floor((endTimestamps[idx] - DateTime.Now).TotalMilliseconds);
			
      try {
				if (sleepTime > 0) {
					Thread.Sleep(sleepTime);
				}
        associatedThreads[idx].canBreak = true;
				idx++;
			} catch (Exception e) {
				Console.WriteLine(e);
			}
		}
	}
}

File: lab1/java/BreakThread.java
import java.util.Arrays;
import java.util.Comparator;
import java.util.Random;
import java.util.stream.Stream;

public class BreakThread extends Thread {
	private static ComputingThreadStatus[] associatedThreads = new ComputingThreadStatus[0];
	
	public BreakThread (int amountThreads) {
    ComputingThread[] threads = new ComputingThread[amountThreads];
    int min = 1;
    int max = 10;
    Random rnd = new Random();

    for (int i = 0; i < amountThreads; i++){
      int step = rnd.nextInt(max - min + 1) + min;
      int runDuration = rnd.nextInt(max - min + 1) + min;
      threads[i] = new ComputingThread(i, 1, step, runDuration);
    }
		ComputingThreadStatus[] threadStatuses = Arrays.stream(threads)
      .map(ComputingThreadStatus::new)
			.sorted(Comparator.comparingInt(ts -> ts.getThread().getRunDuration()))
			.toArray(ComputingThreadStatus[]::new);
		
		for (int i = 0; i < threadStatuses.length; i++) {
			threadStatuses[i].getThread().setQueuePosition(i);
		}
		
		associatedThreads = threadStatuses;
	}
	
	public synchronized static boolean canBreak (int queuePosition) {
		return associatedThreads[queuePosition].isBreak();
	}
	
	@Override
	public void start () {
    long[] endTimestamps = Stream.of(associatedThreads)
    .mapToLong(threadStatus -> System.currentTimeMillis() + threadStatus.getThread().getRunDuration() * 1000)
    .toArray();

		for (ComputingThreadStatus threadStatus : associatedThreads) {
			threadStatus.getThread().start();
		}

		for (int i = 0; i < associatedThreads.length; ) {
			try {
        long sleepTime = (long)(endTimestamps[i] - System.currentTimeMillis());
				if(sleepTime > 0) Thread.sleep(sleepTime);
				associatedThreads[i].setBreak(true);
        i++;
			} catch (Exception e) {
				e.printStackTrace();
        break;
			}
		}
	}
}
